lib/GADS.pm

```diff
@@ -3024,6 +3025,99 @@ prefix '/:layout_name' => sub {
 
     };
 
+    any ['get', 'post'] => '/historic_purge/' => require_login sub {
+
+        my $layout = var('layout') or pass;
+
+        my $user        = logged_in_user;
+        my $view        = current_view($user, $layout);
+
+        my $records = GADS::Records->new(
+            user                => $user,
+            layout              => $layout,
+            schema              => schema,
+            include_children    => 1,
+            view_limit_extra_id => undef, # Override any value that may be set
+            view                => $view,
+        );
+
+        my $params = +{};
+        my $record_data = [];
+
+        if(param 'submit') {
+            my @columns = body_parameters->get_all('record');
+            $records->columns([@columns]);
+        } elsif(param 'purge') {
+            my @columns = split(/,/,body_parameters->get('records'));
+            $records->columns([@columns]);
+        }
+
+        my %ignores = (
+            id => 1,
+            createddate => 1,
+            createdby => 1,
+            serial => 1,
+        );
+
+        unless(param 'purge') {
+            while (my $record = $records->single) {
+                foreach my $column (@{$record->presentation->{columns}}) {
+                    next if $ignores{$column->{type}};
+                    my $result = +{};
+                    $result->{col_id} = $column->{id};
+                    $result->{name} = $column->{name};
+                    $result->{id} = $column->{data}->{record_id};
+                    $result->{value} = $column->{data}->{value};
+                    push @$record_data, $result;
+                }
+            }
+        }
+
+        $params->{page} = 'initial_purge';
+
+        if(param 'submit') {
+            $params->{page} = 'purge_stage_2';
+            $params->{records} = $record_data;
+            my @mapped_ids = map {$_->{col_id}} @$record_data;
+            my @mapped_records = uniq map {$_->{id}} @$record_data;
+            $params->{column_list} = join(',', @mapped_ids);
+            $params->{record_list} = join(',', @mapped_records);
+        } elsif(param 'purge') {
+            my $columns = body_parameters->get('columns') or error __"Unable to get columns";
+            my $rids = body_parameters->get('records') or error __"Unable to get records";
+            my @mapped_ids = split(/,/,$columns);
+            my @records = split(/,/,$rids);
```

Comment:
I'd prefer to use the records from the view (simpler), unless the idea is to prevent records being purged that were not in the view originally?